### PR TITLE
Dmlf client rework

### PR DESCRIPTION
**Pull request by baykaner**

### CODE REVIEW

File: libs/dmlf/include/dmlf/collective_learning/utilities/boston_housing_client_utilities.hpp [hunk 1]
```diff
@@ -17,24 +17,30 @@
...
+namespace {
+/**
+ * Utility for setting up a single MnistModel
+ * @tparam TensorType
+ * @param client_params
+ * @param images
+ * @param labels
+ * @param test_set_ratio
+ * @return
+ */
```
Comment (evsmithx) [inline libs/dmlf/include/dmlf/collective_learning/utilities/boston_housing_client_utilities.hpp:39]: It's not mnist! And the parameter list is incorrect.

File: libs/dmlf/include/dmlf/collective_learning/utilities/mnist_client_utilities.hpp [hunk 2]
```diff
@@ -56,15 +63,53 @@ std::shared_ptr<fetch::dmlf::distributed_learning::TrainingClient<TensorType>> M
...
+ * @tparam TensorType
+ * @param id
+ * @param client_params
+ * @param images
+ * @param labels
+ * @param test_set_ratio
+ * @param networker
+ * @param console_mutex_ptr
+ * @return
+ */
```
Comment (evsmithx) [inline libs/dmlf/include/dmlf/collective_learning/utilities/mnist_client_utilities.hpp:85]: Comment says every algorithm, but this is just for MNIST?

File: libs/dmlf/include/dmlf/collective_learning/client_algorithm.hpp [hunk 3]
```diff
@@ -106,109 +119,75 @@ class TrainingClient
...
   , console_mutex_ptr_(std::move(console_mutex_ptr))
+  , params_(client_params)
+  , algorithm_controller_(algorithm_controller)
 {
-  dataloader_ptr_ = model_ptr_->dataloader_ptr_;
-  graph_ptr_      = model_ptr_->graph_ptr_;
-  optimiser_ptr_  = model_ptr_->optimiser_ptr_;
-
-  SetParams(client_params);
+  //  SetParams(client_params);
```
Comment (evsmithx) [inline libs/dmlf/include/dmlf/collective_learning/client_algorithm.hpp]: commented code

File: libs/dmlf/include/dmlf/collective_learning/client_algorithm.hpp [hunk 4]
```diff
@@ -395,67 +377,58 @@ void TrainingClient<TensorType>::SetWeights(VectorTensorType const &new_weights)
...
-      .count();
-}
-
 /**
- * Do one batch and broadcast gradients
+ * Perform one round of training. This includes
+ * 1. local training
+ * 2. push local updates to algorithm controller
+ * 3. apply updates received by algorithm controller
+ * 4. aggregate and apply updates
```
Comment (evsmithx) [inline libs/dmlf/include/dmlf/collective_learning/client_algorithm.hpp:390]: Helpful comment! +1

File: libs/dmlf/include/dmlf/collective_learning/collective_learning_client.hpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+template <class TensorType>
+template <typename AlgorithmType, typename ClientParamsType>
+void CollectiveLearningClient<TensorType>::BuildAlgorithms(
+    ClientParamsType const &client_params, std::shared_ptr<std::mutex> console_mutex_ptr)
+{
+  for (std::size_t i = 0; i < client_params.n_algorithms_per_client; ++i)
+  {
+    AlgorithmType tmp(algorithm_controller_, "client" + id_ + "_algo" + std::to_string(i),
+                      client_params, console_mutex_ptr);
```
Comment (evsmithx) [inline libs/dmlf/include/dmlf/collective_learning/collective_learning_client.hpp]: Debug code?

**Review by evsmithx — CHANGES_REQUESTED** [review]

File: libs/dmlf/include/dmlf/collective_learning/collective_learning_client.hpp (hunk 5 above)
Comment (baykaner) [reply libs/dmlf/include/dmlf/collective_learning/collective_learning_client.hpp]: yup; forgot to ditch it

File: libs/dmlf/include/dmlf/collective_learning/client_algorithm.hpp (hunk 4 above)
Comment (baykaner) [reply libs/dmlf/include/dmlf/collective_learning/client_algorithm.hpp:390]: a rarity i know :-p

File: libs/dmlf/include/dmlf/collective_learning/utilities/mnist_client_utilities.hpp (hunk 2 above)
Comment (baykaner) [reply libs/dmlf/include/dmlf/collective_learning/utilities/mnist_client_utilities.hpp:85]: it means every algorithm owned by the client. in this case all of the algorithms are mnist algorithms.
The point here is that one client can own multiple algorithms.

**Review by evsmithx — APPROVED** [review]

**Pull request merged**